satellite/rewards/partners_service.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package rewards
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"path"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+)
+
+var (
+	// Error is the default error class for partners package.
+	Error = errs.Class("partners error class")
+
+	// ErrNotExist is returned when a particular partner does not exist.
+	ErrNotExist = errs.Class("partner does not exist")
+)
+
+// PartnersDB allows access to partners database.
+//
+// architecture: Database
+type PartnersDB interface {
+	// All returns all partners.
+	All(ctx context.Context) ([]PartnerInfo, error)
+	// ByName returns partner definitions for a given name.
+	ByName(ctx context.Context, name string) (PartnerInfo, error)
+	// ByID returns partner definition corresponding to an id.
+	ByID(ctx context.Context, id string) (PartnerInfo, error)
+	// ByUserAgent returns partner definition corresponding to an user agent string.
+	ByUserAgent(ctx context.Context, agent string) (PartnerInfo, error)
+}
+
+// PartnersService allows manipulating and accessing partner information.
+//
+// architecture: Service
+type PartnersService struct {
+	log     *zap.Logger
+	db      PartnersDB
+	domains []string
+}
+
+// NewPartnersService returns a service for handling partner information.
+func NewPartnersService(log *zap.Logger, db PartnersDB, domains []string) *PartnersService {
+	return &PartnersService{
+		log:     log,
+		db:      db,
+		domains: domains,
+	}
+}
+
+// GeneratePartnerLink returns base64 encoded partner referral link.
+func (service *PartnersService) GeneratePartnerLink(ctx context.Context, offerName string) ([]string, error) {
+	partner, err := service.db.ByName(ctx, offerName)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	type info struct {
+		UserID    string
+		PartnerID string
+	}
+
+	referralInfo := &info{UserID: "", PartnerID: partner.ID}
+	refJSON, err := json.Marshal(referralInfo)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	// TODO: why is this using base64?
+	encoded := base64.StdEncoding.EncodeToString(refJSON)
+
+	var links []string
+	for _, domain := range service.domains {
+		links = append(links, path.Join(domain, "ref", encoded))
+	}
+
+	return links, nil
+}
+
+// GetActiveOffer returns an offer that is active based on its type
+func (service *PartnersService) GetActiveOffer(ctx context.Context, offers Offers, offerType OfferType, partnerID string) (offer *Offer, err error) {
```

Comment:
I know you said that we want to keep the minimal functional change in this PR. I assume you would change this method signature too in the next PR since now it's in the `PartnerService` ?

Reply:
Yeah. I'm definitely not happy where things are at the moment.